File: ios/SOTALog/ViewModels/SpotsViewModel.swift
import Foundation
import Observation

@Observable
final class SpotsViewModel {
    enum SourceFilter: String {
        case all, pota, sota
    }

    enum ModeFilter: String {
        case all, cw, ssb
    }

    var spots: [Spot] = []
    var sourceFilter: SourceFilter = .all
    var modeFilter: ModeFilter = .all
    var isLoading = false
    var errorMessage: String?

    private var sotaEpoch: String?
    private var sotaSpots: [Spot] = []
    private var potaSpots: [Spot] = []

    init() {
        if let raw = UserDefaults.standard.string(forKey: "sourceFilter"),
           let saved = SourceFilter(rawValue: raw) {
            sourceFilter = saved
        }
        if let raw = UserDefaults.standard.string(forKey: "modeFilter"),
           let saved = ModeFilter(rawValue: raw) {
            modeFilter = saved
        }
    }

    /// Spots consolidated (one per callsign), grouped by band, sorted by frequency then time.
    var spotsByBand: [(band: String, spots: [Spot])] {
        let consolidated = consolidatedSpots(spots).filter { !$0.isQRT && !$0.isExpired() }

        let sourceFiltered: [Spot]
        switch sourceFilter {
        case .all:  sourceFiltered = consolidated
        case .pota: sourceFiltered = consolidated.filter { $0.potaReference != nil }
        case .sota: sourceFiltered = consolidated.filter { $0.sotaReference != nil }
        }

        let filtered: [Spot]
        switch modeFilter {
        case .all: filtered = sourceFiltered
        case .cw:  filtered = sourceFiltered.filter { $0.mode == "CW" }
        case .ssb: filtered = sourceFiltered.filter { $0.mode == "SSB" }
        }

        // Group by band
        var grouped: [String: [Spot]] = [:]
        for spot in filtered {
            grouped[spot.band, default: []].append(spot)
        }

        // Sort spots within each band by frequency ascending, then newest first
        for (band, bandSpots) in grouped {
            grouped[band] = bandSpots.sorted { a, b in
                if a.frequency != b.frequency { return a.frequency < b.frequency }
                return a.timestamp > b.timestamp
            }
        }

        // Order bands by BandPlan order
        return BandPlan.allBands.compactMap { band in
            guard let bandSpots = grouped[band], !bandSpots.isEmpty else { return nil }
            return (band: band, spots: bandSpots)
        }
    }

    func refresh() async {
        isLoading = true
        defer { isLoading = false }

        do {
            async let pota = POTASpotService.fetchSpots()
            async let sota = SOTASpotService.fetchSpots()

            let (potaResult, sotaResult) = try await (pota, sota)
            potaSpots = potaResult
            sotaSpots = sotaResult
            sotaEpoch = nil  // Reset epoch so next poll fetches fresh
            spots = potaSpots + sotaSpots
            errorMessage = nil
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func startAutoRefresh() async {
        // Initial full fetch of both sources
        await refresh()

        var tickCount = 0

        // Poll every 20 seconds
        while !Task.isCancelled {
            try? await Task.sleep(for: .seconds(20))
            guard !Task.isCancelled else { break }

            tickCount += 1

            do {
                // Every tick: check SOTA epoch, fetch only if changed
                let epoch = try await SOTASpotService.fetchEpoch()
                if epoch != sotaEpoch {
                    sotaSpots = try await SOTASpotService.fetchSpots()
                    sotaEpoch = epoch
                }

                // Every 3rd tick (~60s): also fetch POTA
                if tickCount % 3 == 0 {
                    potaSpots = try await POTASpotService.fetchSpots()
                }

                spots = potaSpots + sotaSpots
                errorMessage = nil
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    /// Find a spot for a given callsign (for auto-populating QSO entry)
    func spotForCallsign(_ callsign: String) -> Spot? {
        let consolidated = consolidatedSpots(spots)
        return consolidated.first { $0.activatorCallsign.uppercased() == callsign.uppercased() && !$0.isExpired() && !$0.isQRT }
    }

    /// Consolidates spots: latest per callsign, merging POTA+SOTA references for same callsign.
    private func consolidatedSpots(_ allSpots: [Spot]) -> [Spot] {
        var byCallsign: [String: Spot] = [:]

        // Sort by timestamp descending so first encounter per callsign is the newest
        let sorted = allSpots.sorted { $0.timestamp > $1.timestamp }

        for spot in sorted {
            let key = spot.activatorCallsign.uppercased()

            if var existing = byCallsign[key] {
                // Merge references from this spot into the existing (newer) one
                if existing.potaReference == nil, let ref = spot.potaReference {
                    existing.potaReference = ref
                    existing.potaReferenceName = spot.potaReferenceName
                }
                if existing.sotaReference == nil, let ref = spot.sotaReference {
                    existing.sotaReference = ref
                    existing.sotaReferenceName = spot.sotaReferenceName
                }
                byCallsign[key] = existing
            } else {
                byCallsign[key] = spot
            }
        }

        return Array(byCallsign.values)
    }
}

File: android/app/src/main/kotlin/com/sotalog/android/ui/spots/SpotsViewModel.kt
package com.sotalog.android.ui.spots

import android.content.SharedPreferences
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sotalog.android.data.remote.api.POTASpotApi
import com.sotalog.android.data.remote.api.SOTASpotApi
import com.sotalog.android.domain.models.Spot
import com.sotalog.android.domain.services.BandPlan
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import javax.inject.Inject

enum class SourceFilter(val value: String) {
    ALL("all"),
    POTA("pota"),
    SOTA("sota"),
}

enum class ModeFilter(val value: String) {
    ALL("all"),
    CW("cw"),
    SSB("ssb"),
}

@HiltViewModel
class SpotsViewModel @Inject constructor(
    private val potaSpotApi: POTASpotApi,
    private val sotaSpotApi: SOTASpotApi,
    private val json: Json,
) : ViewModel() {

    private val _spots = MutableStateFlow<List<Spot>>(emptyList())
    val spots: StateFlow<List<Spot>> = _spots.asStateFlow()

    private val _sourceFilter = MutableStateFlow(SourceFilter.ALL)
    val sourceFilter: StateFlow<SourceFilter> = _sourceFilter.asStateFlow()

    private val _modeFilter = MutableStateFlow(ModeFilter.ALL)
    val modeFilter: StateFlow<ModeFilter> = _modeFilter.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private var potaSpots: List<Spot> = emptyList()
    private var sotaSpots: List<Spot> = emptyList()
    private var autoRefreshJob: Job? = null
    private var sotaEpoch: String? = null

    data class SpotBandGroup(
        val band: String,
        val spots: List<Spot>,
    )

    val spotsByBand: StateFlow<List<SpotBandGroup>> = combine(
        _spots, _sourceFilter, _modeFilter,
    ) { spots, source, mode ->
        val consolidated = consolidateSpots(spots)
        val afterFilter = consolidated.filter { !it.isQRT && !it.isExpired() }

        val sourceFiltered = when (source) {
            SourceFilter.ALL -> afterFilter
            SourceFilter.POTA -> afterFilter.filter { it.potaReference != null }
            SourceFilter.SOTA -> afterFilter.filter { it.sotaReference != null }
        }

        val filtered = when (mode) {
            ModeFilter.ALL -> sourceFiltered
            ModeFilter.CW -> sourceFiltered.filter { it.mode == "CW" }
            ModeFilter.SSB -> sourceFiltered.filter { it.mode == "SSB" }
        }

        val grouped = mutableMapOf<String, MutableList<Spot>>()
        for (spot in filtered) {
            grouped.getOrPut(spot.band) { mutableListOf() }.add(spot)
        }

        for ((band, bandSpots) in grouped) {
            grouped[band] = bandSpots.sortedWith(
                compareBy<Spot> { it.frequency }.thenByDescending { it.timestamp },
            ).toMutableList()
        }

        BandPlan.allBands.mapNotNull { band ->
            val bandSpots = grouped[band]
            if (bandSpots.isNullOrEmpty()) null
            else SpotBandGroup(band = band, spots = bandSpots)
        }
    }.stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    fun setSourceFilter(filter: SourceFilter) {
        _sourceFilter.value = filter
    }

    fun setModeFilter(filter: ModeFilter) {
        _modeFilter.value = filter
    }

    fun refresh() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                val potaResult = try { parsePOTASpots(potaSpotApi.getActivatorSpots().string()) } catch (_: Exception) { emptyList() }
                val sotaResult = try { parseSOTASpots(sotaSpotApi.getSpots().string()) } catch (_: Exception) { emptyList() }
                potaSpots = potaResult
                sotaSpots = sotaResult
                sotaEpoch = null // Reset epoch so next poll fetches fresh
                _spots.value = potaSpots + sotaSpots
                _errorMessage.value = null
            } catch (e: Exception) {
                _errorMessage.value = e.message
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun startAutoRefresh() {
        autoRefreshJob?.cancel()
        autoRefreshJob = viewModelScope.launch {
            refresh()

            var tickCount = 0
            while (isActive) {
                delay(20_000)
                tickCount++

                try {
                    // Check SOTA epoch — only refetch if spots have changed
                    val epoch = try { sotaSpotApi.getEpoch().string().trim('"', ' ') } catch (_: Exception) { null }
                    if (epoch != null && epoch != sotaEpoch) {
                        sotaSpots = try { parseSOTASpots(sotaSpotApi.getSpots().string()) } catch (_: Exception) { sotaSpots }
                        sotaEpoch = epoch
                    }

                    // Every 3rd tick (~60s): also fetch POTA
                    if (tickCount % 3 == 0) {
                        potaSpots = try { parsePOTASpots(potaSpotApi.getActivatorSpots().string()) } catch (_: Exception) { potaSpots }
                    }

                    _spots.value = potaSpots + sotaSpots
                    _errorMessage.value = null
                } catch (e: Exception) {
                    _errorMessage.value = e.message
                }
            }
        }
    }

    fun stopAutoRefresh() {
        autoRefreshJob?.cancel()
        autoRefreshJob = null
    }

    internal fun setSpots(spots: List<Spot>) {
        _spots.value = spots
    }

    fun spotForCallsign(callsign: String): Spot? {
        val consolidated = consolidateSpots(_spots.value)
        return consolidated.firstOrNull {
            it.activatorCallsign.equals(callsign, ignoreCase = true) && !it.isExpired() && !it.isQRT
        }
    }

    private fun consolidateSpots(allSpots: List<Spot>): List<Spot> {
        val byCallsign = linkedMapOf<String, Spot>()
        val sorted = allSpots.sortedByDescending { it.timestamp }

        for (spot in sorted) {
            val key = spot.activatorCallsign.uppercase()
            val existing = byCallsign[key]
            if (existing != null) {
                var merged = existing
                if (merged.potaReference == null && spot.potaReference != null) {
                    merged = merged.copy(
                        potaReference = spot.potaReference,
                        potaReferenceName = spot.potaReferenceName,
                    )
                }
                if (merged.sotaReference == null && spot.sotaReference != null) {
                    merged = merged.copy(
                        sotaReference = spot.sotaReference,
                        sotaReferenceName = spot.sotaReferenceName,
                    )
                }
                byCallsign[key] = merged
            } else {
                byCallsign[key] = spot
            }
        }

        return byCallsign.values.toList()
    }

    // MARK: - Spot Parsing

    private fun parsePOTASpots(jsonStr: String): List<Spot> {
        val array = try { json.parseToJsonElement(jsonStr).jsonArray } catch (_: Exception) { return emptyList() }
        return array.mapNotNull { element ->
            try {
                val obj = element.jsonObject
                val callsign = obj["activator"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
                val freq = obj["frequency"]?.jsonPrimitive?.contentOrNull?.toDoubleOrNull()?.let { it / 1000.0 } ?: return@mapNotNull null
                val mode = obj["mode"]?.jsonPrimitive?.contentOrNull?.uppercase() ?: "CW"
                val reference = obj["reference"]?.jsonPrimitive?.contentOrNull
                val refName = obj["name"]?.jsonPrimitive?.contentOrNull
                val spotter = obj["spotter"]?.jsonPrimitive?.contentOrNull
                val comments = obj["comments"]?.jsonPrimitive?.contentOrNull
                val timestamp = obj["spotTime"]?.jsonPrimitive?.contentOrNull?.let { parseTimestamp(it) } ?: Date()

                val normalizedMode = when {
                    mode.contains("CW") -> "CW"
                    mode.contains("SSB") || mode.contains("LSB") || mode.contains("USB") || mode.contains("PHONE") -> "SSB"
                    mode.contains("FT") || mode.contains("DIGI") -> "FT8"
                    else -> mode
                }

                Spot(
                    id = "pota_${callsign}_${reference}",
                    activatorCallsign = callsign,
                    frequency = freq,
                    mode = normalizedMode,
                    potaReference = reference,
                    potaReferenceName = refName,
                    spotterCallsign = spotter,
                    comments = comments,
                    timestamp = timestamp,
                )
            } catch (_: Exception) {
                null
            }
        }
    }

    private fun parseSOTASpots(jsonStr: String): List<Spot> {
        val array = try { json.parseToJsonElement(jsonStr).jsonArray } catch (_: Exception) { return emptyList() }
        return array.mapNotNull { element ->
            try {
                val obj = element.jsonObject
                val callsign = obj["activatorCallsign"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
                val freq = obj["frequency"]?.jsonPrimitive?.doubleOrNull ?: return@mapNotNull null
                if (freq <= 0.0) return@mapNotNull null
                val mode = obj["mode"]?.jsonPrimitive?.contentOrNull?.uppercase() ?: "CW"
                val reference = obj["summitCode"]?.jsonPrimitive?.contentOrNull
                val refName = obj["summitName"]?.jsonPrimitive?.contentOrNull
                val spotter = obj["callsign"]?.jsonPrimitive?.contentOrNull
                val rawComments = obj["comments"]?.jsonPrimitive?.contentOrNull
                val type = obj["type"]?.jsonPrimitive?.contentOrNull
                val comments = if (type?.uppercase() == "QRT") {
                    val existing = rawComments ?: ""
                    if (existing.uppercase().contains("QRT")) existing else "QRT $existing".trim()
                } else {
                    rawComments
                }
                val spotId = obj["id"]?.jsonPrimitive?.contentOrNull ?: "0"
                val timestamp = obj["timeStamp"]?.jsonPrimitive?.contentOrNull?.let { parseTimestamp(it) } ?: Date()

                val normalizedMode = when {
                    mode.contains("CW") -> "CW"
                    mode.contains("SSB") || mode.contains("LSB") || mode.contains("USB") || mode.contains("FM") -> "SSB"
                    mode.contains("FT") || mode.contains("DATA") -> "FT8"
                    else -> mode
                }

                Spot(
                    id = "sota-${spotId}-${callsign}",
                    activatorCallsign = callsign,
                    frequency = freq,
                    mode = normalizedMode,
                    sotaReference = reference,
                    sotaReferenceName = refName,
                    spotterCallsign = spotter,
                    comments = comments,
                    timestamp = timestamp,
                )
            } catch (_: Exception) {
                null
            }
        }
    }

    private fun parseTimestamp(str: String): Date? {
        val formats = listOf(
            "yyyy-MM-dd'T'HH:mm:ss",
            "yyyy-MM-dd'T'HH:mm:ssX",
            "yyyy-MM-dd'T'HH:mm:ss.SSSX",
            "yyyy-MM-dd HH:mm:ss",
        )
        for (format in formats) {
            try {
                val sdf = SimpleDateFormat(format, Locale.US)
                sdf.timeZone = TimeZone.getTimeZone("UTC")
                return sdf.parse(str)
            } catch (_: Exception) {
                // try next format
            }
        }
        return null
    }
}
